# Spam Detector

In [1]:
import  pandas                  as pd
from    sklearn.model_selection import train_test_split
from    sklearn.metrics         import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

** I predict Logistic Regression will perform best.  Since the model needs to classify the data as either spam or not spam, this is a binary classification and Linear Regression is great for binary classification.  

## Split the Data into Training and Testing Sets

In [4]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]
#X = df.copy()
X = data.drop(columns='spam')

In [6]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [13]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [14]:
display(X_train.head())
display(X_test.head())

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1570            0.00               0.00           1.07           0.0   
1673            0.13               0.10           0.55           0.0   
275             0.00               0.17           0.00           0.0   
3295            0.00               0.00           1.06           0.0   
233             0.00               0.00           0.72           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1570           3.22            0.00              0.00                0.00   
1673           0.02            0.13              0.02                1.11   
275            0.00            0.00              0.17                0.52   
3295           0.00            0.00              0.00                0.00   
233            1.81            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1570             0.00            0.00  ...                   0.0        0.000   
1673             0.23            0.29  ...                   0.0        0.030   
275              0.00            0.17  ...                   0.0        0.028   
3295             0.00            0.00  ...                   0.0        0.000   
233              0.00            0.36  ...                   0.0        0.053   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1570        0.000        0.000        2.395        0.598        0.998   
1673        0.083        0.000        0.538        0.145        0.070   
275         0.143        0.028        0.085        0.057        0.229   
3295        0.121        0.000        0.000        0.000        0.000   
233         0.265        0.000        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1570                      82.250                         295   
1673                       5.108                         116   
275                        3.564                          39   
3295                       2.151                          10   
233                        1.543                          13   

      capital_run_length_total  
1570                       329  
1673                      3525  
275                        417  
3295                        71  
233                         88  

[5 rows x 57 columns]

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
4572            0.00               0.00           0.00           0.0   
1742            0.19               0.19           0.29           0.0   
1806            0.10               0.20           1.01           0.0   
2176            0.00               0.00           0.00           0.0   
938             0.05               0.05           0.40           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
4572           0.00            0.00              0.00                0.00   
1742           1.07            0.19              0.19                0.97   
1806           0.80            0.80              0.50                0.00   
2176           0.00            0.00              0.00                0.00   
938            0.34            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
4572             0.00            0.00  ...                   0.0        0.000   
1742             0.87            0.58  ...                   0.0        0.000   
1806             0.80            0.10  ...                   0.0        0.000   
2176             0.00            0.00  ...                   0.0        0.000   
938              0.57            0.05  ...                   0.0        0.019   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
4572        0.254        0.000        0.000        0.000        0.000   
1742        0.000        0.202        0.404        0.233        0.031   
1806        0.110        0.000        0.488        0.157        0.015   
2176        0.000        0.000        0.000        0.000        0.000   
938         0.099        0.000        0.089        0.079        0.009   

      capital_run_length_average  capital_run_length_longest  \
4572                       1.000                           1   
1742                       4.320                          49   
1806                       8.550                         669   
2176                       1.400                           3   
938                        4.913                          95   

      capital_run_length_total  
4572                        13  
1742                       877  
1806                      1351  
2176                         7  
938                       1312  

[5 rows x 57 columns]

In [15]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [16]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [27]:
# Scale the training data
X_train_scaled  = scaler.transform(X_train)
X_test_scaled   = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [29]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=1)
logistic_regression_model.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [32]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_regression_model.predict(X_test_scaled)

# Review the predictions
display(testing_predictions)

array([0, 1, 1, ..., 1, 1, 0], dtype=int64)

In [33]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)

0.9287576020851434

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [34]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

rfc_model = RandomForestClassifier(random_state=1)
rfc_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=1)

In [35]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = rfc_model.predict(X_test_scaled)

# Review the predictions
display(testing_predictions)

array([0, 1, 1, ..., 1, 1, 0], dtype=int64)

In [36]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)

0.952215464813206

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forecast Classifer model performed best by a slight margin.  While my prediction was wrong, the Logistic Regression model performed extremely well with around a 93% accuracy compared to the RFC's 95.2%.  I would feel confident using either model, if this were a real world scenario.

In [ ]:
#predictions_df = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_train})
#predictions_df